lightning/src/routing/router.rs

```diff
@@ -333,79 +549,229 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 		};
 	}
 
-	match network.get_nodes().get(target) {
-		None => {},
-		Some(node) => {
-			add_entries_to_cheapest_to_target_node!(node, target, 0);
-		},
-	}
+	let mut payment_paths = Vec::<PaymentPath>::new();
+
+	// TODO: diversify by nodes (so that all paths aren't doomed if one node is offline).
+	'paths_collection: loop {
+		// For every new path, start from scratch, except bookkeeped_channels_liquidity_available_msat,
+		// which will improve the further iterations of path finding. Also don't erase first_hop_targets.
+		targets.clear();
+		dist.clear();
+
+		// Add the payment receiver node as a target, so that the dest-to-source search algorithm knows what to start with.
+		match network.get_nodes().get(target) {
+			None => {},
+			Some(node) => {
+				add_entries_to_cheapest_to_target_node!(node, target, 0);
+			},
+		}
 
-	for hop in last_hops.iter() {
-		if first_hops.is_none() || hop.src_node_id != *our_node_id { // first_hop overrules last_hops
-			if network.get_nodes().get(&hop.src_node_id).is_some() {
-				if first_hops.is_some() {
-					if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
-						// Currently there are no channel-context features defined, so we are a
-						// bit lazy here. In the future, we should pull them out via our
-						// ChannelManager, but there's no reason to waste the space until we
-						// need them.
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, features.to_context(), 0);
+		// If a caller provided us with last hops, add them to routing targets.
+		// Since this happens earlier than general path finding, they will be somewhat prioritized,
+		// although currently it matters only if the fees are exactly the same.
+		for hop in last_hops.iter() {
+			if first_hops.is_none() || hop.src_node_id != *our_node_id { // first_hop overrules last_hops
+				if network.get_nodes().get(&hop.src_node_id).is_some() {
+					if first_hops.is_some() {
+						if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
+							// Currently there are no channel-context features defined, so we are a
+							// bit lazy here. In the future, we should pull them out via our
+							// ChannelManager, but there's no reason to waste the space until we
+							// need them.
+							add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, None::<u64>, features.to_context(), 0);
+						}
 					}
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+					// really sucks, cause we're gonna need that eventually.
```

Comment:
This comment let me wonder if we could do all channel features/routing policy sanitization out of the pure routing logic in the future. Our features-of-interest aren't function of the value transferred, so I don't think that's a blocker to move those checks at gossips reception ?

Reply:
This is part of the legacy code which I copypasted, perhaps leave it for future improvement?

Reply:
Added to backlog.